### ISSUE TITLE
learning rate does not change although arch_improvement_threshold has reached

### ISSUE THREAD

**Issue by alan101-tech** [mravanelli/pytorch-kaldi#162]
The learning rate does not change although arch_improvement threshold has reached. It seems a bug in pytorch_kaldi as function _is_first_validation(ck, N_ck_tr, config) is to check whether ck is the first validation within chunk, not the whole epoches. If nr_of_valid_per_epoch is set 1,  _is_first_validation will always return True. In this way, lr will not change at all.

**Comment by TParcollet** [comment]
Hey ! Thank you for this remark, could you past the content of a res.res file highlighting this behavior?

**Comment by alan101-tech** [comment]
![1](https://user-images.githubusercontent.com/54306776/64514447-35b3d100-d31d-11e9-93af-32e46b602fd0.png)
This is my res.res file, the lr is always the same as the initial one. Thanks!

**Comment by alan101-tech** [comment]
png not uploaded! Below is the content of file res.res: 

ep=00 tr=[u'xatx_0620_neg_rever_train'] loss=0.251 err=0.095 valid=xatx_0620_neg_rever_dev loss=0.248 err=0.092 lr_architecture1=0.08 time(s)=8296
ep=01 tr=[u'xatx_0620_neg_rever_train'] loss=0.234 err=0.088 valid=xatx_0620_neg_rever_dev loss=0.240 err=0.089 lr_architecture1=0.08 time(s)=7989
ep=02 tr=[u'xatx_0620_neg_rever_train'] loss=0.230 err=0.087 valid=xatx_0620_neg_rever_dev loss=0.238 err=0.088 lr_architecture1=0.08 time(s)=8961
ep=03 tr=[u'xatx_0620_neg_rever_train'] loss=0.228 err=0.086 valid=xatx_0620_neg_rever_dev loss=0.234 err=0.087 lr_architecture1=0.08 time(s)=7928
ep=04 tr=[u'xatx_0620_neg_rever_train'] loss=0.227 err=0.086 valid=xatx_0620_neg_rever_dev loss=0.233 err=0.086 lr_architecture1=0.08 time(s)=7885
ep=05 tr=[u'xatx_0620_neg_rever_train'] loss=0.226 err=0.085 valid=xatx_0620_neg_rever_dev loss=0.232 err=0.086 lr_architecture1=0.08 time(s)=7718
ep=06 tr=[u'xatx_0620_neg_rever_train'] loss=0.225 err=0.085 valid=xatx_0620_neg_rever_dev loss=0.232 err=0.086 lr_architecture1=0.08 time(s)=7959
ep=07 tr=[u'xatx_0620_neg_rever_train'] loss=0.225 err=0.085 valid=xatx_0620_neg_rever_dev loss=0.229 err=0.085 lr_architecture1=0.08 time(s)=10190
ep=08 tr=[u'xatx_0620_neg_rever_train'] loss=0.224 err=0.085 valid=xatx_0620_neg_rever_dev loss=0.230 err=0.085 lr_architecture1=0.08 time(s)=8492
ep=09 tr=[u'xatx_0620_neg_rever_train'] loss=0.224 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.230 err=0.085 lr_architecture1=0.08 time(s)=7641
ep=10 tr=[u'xatx_0620_neg_rever_train'] loss=0.223 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.228 err=0.084 lr_architecture1=0.08 time(s)=7950
ep=11 tr=[u'xatx_0620_neg_rever_train'] loss=0.223 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.227 err=0.084 lr_architecture1=0.08 time(s)=7787
ep=12 tr=[u'xatx_0620_neg_rever_train'] loss=0.223 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.226 err=0.084 lr_architecture1=0.08 time(s)=8638
ep=13 tr=[u'xatx_0620_neg_rever_train'] loss=0.223 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.228 err=0.084 lr_architecture1=0.08 time(s)=7982
ep=14 tr=[u'xatx_0620_neg_rever_train'] loss=0.222 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.228 err=0.085 lr_architecture1=0.08 time(s)=8544
ep=15 tr=[u'xatx_0620_neg_rever_train'] loss=0.222 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.227 err=0.084 lr_architecture1=0.08 time(s)=8803
ep=16 tr=[u'xatx_0620_neg_rever_train'] loss=0.222 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.227 err=0.084 lr_architecture1=0.08 time(s)=8491
ep=17 tr=[u'xatx_0620_neg_rever_train'] loss=0.222 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.225 err=0.084 lr_architecture1=0.08 time(s)=8637
ep=18 tr=[u'xatx_0620_neg_rever_train'] loss=0.222 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.229 err=0.085 lr_architecture1=0.08 time(s)=8793
ep=19 tr=[u'xatx_0620_neg_rever_train'] loss=0.222 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.225 err=0.083 lr_architecture1=0.08 time(s)=8582
ep=20 tr=[u'xatx_0620_neg_rever_train'] loss=0.221 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.226 err=0.084 lr_architecture1=0.08 time(s)=9201
ep=21 tr=[u'xatx_0620_neg_rever_train'] loss=0.221 err=0.084 valid=xatx_0620_neg_rever_dev loss=0.225 err=0.083 lr_architecture1=0.08 time(s)=8712
ep=22 tr=[u'xatx_0620_neg_rever_train'] loss=0.221 err=0.083 valid=xatx_0620_neg_rever_dev loss=0.224 err=0.083 lr_architecture1=0.08 time(s)=9380
ep=23 tr=[u'xatx_0620_neg_rever_train'] loss=0.221 err=0.083 valid=xatx_0620_neg_rever_dev loss=0.226 err=0.084 lr_architecture1=0.08 time(s)=8798

**Comment by TParcollet** [comment]
Hum, could you please past your cfg file ? I'll take a look into the code.

**Comment by alan101-tech** [comment]
hi,  cfg file is :

[cfg_proto]
cfg_proto=proto/global.proto
cfg_proto_chunk=proto/global_chunk.proto

[exp]
cmd=
run_nn_script=run_nn
out_folder=exp/wakeup_MLP_fbank_3x256_2gpu
seed=1234
use_cuda=True
multi_gpu=True
save_gpumem=False
N_epochs_tr=24

[dataset1]
data_name=xatx_0620_neg_rever_train
fea:fea_name=fbank
    fea_lst=/data_ssd/siyujing/git/pytorch-kaldi/data/feats_train.scp
    fea_opts=
    cw_left=25
    cw_right=5

    
lab:lab_name=lab_cd
    lab_folder=/data_ssd/siyujing/git/pytorch-kaldi/data/
    lab_opts=
    lab_count_file=auto
    lab_data_folder=/data4/siyujing/OPENSOURCE/kaldi/egs/librispeech/s5/fmllr/train_clean_100/
    lab_graph=/data4/siyujing/OPENSOURCE/kaldi/egs/librispeech/s5/exp/tri4b/graph_tgsmall/

N_chunks=600
        
[dataset2]
data_name=xatx_0620_neg_rever_dev
fea:fea_name=fbank
    fea_lst=/data_ssd/siyujing/git/pytorch-kaldi/data/feats_dev.scp
    fea_opts=
    cw_left=25
    cw_right=5

lab:lab_name=lab_cd
    lab_folder=/data_ssd/siyujing/git/pytorch-kaldi/data/
    lab_opts=
    lab_count_file=auto
    lab_data_folder=/data4/siyujing/OPENSOURCE/kaldi/egs/librispeech/s5/fmllr/dev_clean/
    lab_graph=/data4/siyujing/OPENSOURCE/kaldi/egs/librispeech/s5/exp/tri4b/graph_tgsmall/

N_chunks=30

[dataset3]
data_name=xatx_0620_neg_rever_test
fea:fea_name=fbank
    fea_lst=/data_ssd/siyujing/git/pytorch-kaldi/data/feats_dev.scp
    fea_opts=
    cw_left=25
    cw_right=5


lab:lab_name=lab_cd
    lab_folder=/data4/siyujing/git/pytorch-kaldi/wakeup_data/
    lab_opts=
    lab_count_file=auto
    lab_data_folder=/data4/siyujing/OPENSOURCE/kaldi/egs/librispeech/s5/fmllr/test_clean/
    lab_graph=/data4/siyujing/OPENSOURCE/kaldi/egs/librispeech/s5/exp/tri4b/graph_tgsmall/

N_chunks=30

        
[data_use]
train_with=xatx_0620_neg_rever_train
valid_with=xatx_0620_neg_rever_dev
forward_with=xatx_0620_neg_rever_test


[batches]
batch_size_train=1024
max_seq_length_train=1000
increase_seq_length_train=False
start_seq_len_train=100
multply_factor_seq_len_train=2
batch_size_valid=4096
max_seq_length_valid=1000


[architecture1]
arch_name=MLP_layers
arch_proto=proto/MLP.proto
arch_library=neural_networks
arch_class=MLP
arch_pretrain_file=none
arch_freeze=False
arch_seq_model=False

dnn_lay=256,256,256,6
dnn_drop=0.0,0.0,0.0,0.0
dnn_use_laynorm_inp=False
dnn_use_batchnorm_inp=False
dnn_use_batchnorm=True,True,True,False
dnn_use_laynorm=False,False,False,False
dnn_act=relu,relu,relu,softmax

arch_lr=0.08
arch_halving_factor=0.5
arch_improvement_threshold=0.001
arch_opt=sgd
opt_momentum=0.0
opt_weight_decay=0.0
opt_dampening=0.0
opt_nesterov=False


[model]
model_proto=proto/model.proto
model:out_dnn1=compute(MLP_layers,fbank)
      loss_final=cost_nll(out_dnn1,lab_cd)
      err_final=cost_err(out_dnn1,lab_cd)
 
      
[forward]
forward_out=out_dnn1
normalize_posteriors=False
normalize_with_counts_from=lab_cd
save_out_file=False
require_decoding=False


[decoding]
decoding_script_folder=kaldi_decoding_scripts/
decoding_script=decode_dnn.sh
decoding_proto=proto/decoding.proto
min_active=200
max_active=7000
max_mem=50000000
beam=20.0
latbeam=12.0
acwt=0.10
max_arcs=-1
skip_scoring=false
scoring_script=/data4/siyujing/OPENSOURCE/kaldi/egs/librispeech/s5/local/score.sh
scoring_opts="--min-lmwt 4 --max-lmwt 23"
norm_vars=False

**Comment by TParcollet** [comment]
Ok. your threshold is 0.001. The LR is reduced when (err_prev - err_curent)/err_current is < to 0.001. Unfortunately, in your case it's 0.0011 which is higher. Try increasing this factor to 0.01 or even higher. You could also directly add a LR schedule as specified in the README :)

**Comment by alan101-tech** [comment]
but the error of epoch 8 and epoch 9 are both 0.085, why lr does not change?

**Comment by TParcollet** [comment]
@alan101-tech True, sorry. We are looking into that. You could try to set your own LR schedule for now?

**Comment by TParcollet** [comment]
@mennetob any idea?

**Comment by mennetob** [comment]
I think @alan101-tech is right. `_is_first_validation` should be made dependent on `ep` and return `False` if `ep > 0`

@alan101-tech have you tried to set `nr_of_valid_per_epoch` to e.g. 5 just to validate whether the lr is adapted then? Just to make sure that there is not another thing going wrong.

**Comment by alan101-tech** [comment]
yeah, i set nr_of_valid_per_epoch to 2, and the lr can change.

**Comment by TParcollet** [comment]
@mennetob could you provide a fix on this? It will affect all the experiments where this variable is not set correctly.

**Comment by mennetob** [comment]
Oh yes sure I can do that.
I'm currently out of office. But I can have it done by the end of the week.

**Comment by mravanelli** [comment]
I fixed it. Unfortunately this bug might have affected many users (it is my fault, I should have reviewed more carefully the past pull requests). @mennetob do you know if this issue has affected some of the CHiME 5 experiments as well? When you have time, it could make sense to comment a bit more the parts in run_exp.py that you added during JSALT (we want to maintain this code until a first version of SpeechBrain is out).
